README.md
[![Lint with flake8](https://github.com/apider/whale-alert-es-pub/actions/workflows/flake8-lint.yml/badge.svg)](https://github.com/apider/whale-alert-es-pub/actions/workflows/flake8-lint.yml)
[![CodeQL](https://github.com/apider/whale-alert-es-pub/actions/workflows/github-code-scanning/codeql/badge.svg)](https://github.com/apider/whale-alert-es-pub/actions/workflows/github-code-scanning/codeql)
[![CI Pipeline](https://github.com/apider/whale-alert-es-pub/actions/workflows/ci.yaml/badge.svg)](https://github.com/apider/whale-alert-es-pub/actions/workflows/ci.yaml)
[![Release](https://github.com/apider/whale-alert-es-pub/actions/workflows/release.yaml/badge.svg)](https://github.com/apider/whale-alert-es-pub/actions/workflows/release.yaml)

# whale-alert-es

Whale Alert crypto transaction scraper to Elasticsearch

Dockerhub: https://hub.docker.com/r/apider/whale-alert-es-pub

Scrapes whale-alert crypto transaction data and ingests to Elasticsearch for further dashboarding/graphing.

This project was done for learning more about async functionality and as a fun crypto project.

Elasticsearch ingest is done asynchronously with asyncio & httpx library.
### Uses data from
Main site:  
https://whale-alert.io/

Docs: 
https://docs.whale-alert.io/

## Config ENV vars
Runtime ENV vars (mandator/optional):

<code>TOKEN</code> - your personal Whale Alert API key - mandatory

<code>CURSORPATH</code> - path for <code>cursor</code> file, default <code>/data/</code> - optional

<code>MIN_VALUE</code> - minimum transaction value in $, default <code>500000</code> - optional

<code>ESHOST</code> - Elasticsearch host, default <code>http://elasticsearch:9200</code> - optional

<code>ESINDEX</code> - Elasticsearch index name, default <code>whalealert</code> - optional

### API-key / TOKEN
Register with whale-alert and get your private api key. https://whale-alert.io/

You can store api key in a file if you want. 
Change line <code>11-13</code> in <code>app.py</code> accordingly.

### cursor
Whale Alert API gives you a cursor so you can keep track of where in the event flow you are.

I save this cursor in a file called <code>cursor.json</code> to be loaded on program/container restarts so we dont re-fetch old data.

If the cursor is older than 3600 seconds, API throws an error and the code handles that and removes the outdated/incorrect cursor.

The <code>manifest.yaml</code> has a PVC & mount defined for this.

If you run pure <code>docker</code> you need to create & use a volume for it.

### Elasticsearch config
See 'Config ENV vars' above.

# Running
## locally
Just run <code>python3 app.py</code> on your workstation/server.

Remember to set the <code>TOKEN</code> and optionally <code>CURSORPATH</code> ENV vars (see above).
## Docker
Use <code>Dockerfile</code> to build from, if you want to run in container or Kubernetes.

You will need to create & use a volume for <code>cursorpath</code> for pure docker.

## Kubernetes
There is a basic <code>manifest.yaml</code> containing Deployment & PVC.
### Name space
The <code>manifest.yaml</code> uses name space <code>prod</code>
### api-key in k8s
Before deploying in Kubernetes, first create a secret named <code>token</code>

Deployment will import it as ENV variable <code>TOKEN</code>

<code>kubectl -n YOUR-NAME-SPACE create secret generic whale-alert-es-secret --from-literal=token='your-whale-alert-api-key'</code>

### k8s image repository
You need to change the repository url & port in <code>manifest.yaml</code> on line <code>29</code>, as manifest uses my local repo.

## Dashboard
A simple Kibana dashboard & viz export can be found in <code>kibana/kibana-dashboard-and-viz-export.json</code>

This dash could be done much better ofcourse. Also I have a really old ES/Kibana version.

![alt text](https://github.com/apider/whale-alert-es/blob/master/kibana/dashboard.png?raw=true)
# whale-alert-priv


CHANGELOG.md
## [0.1.1](https://github.com/apider/whale-alert-es-pub/compare/v0.1.0...v0.1.1) (2023-05-29)


### Bug Fixes

* new release ([6f76756](https://github.com/apider/whale-alert-es-pub/commit/6f767567fe6d996cee91744d7bf0eff90a2e2083))


app.py
import json
import asyncio
import time
import logging
import os
# import socket
from datetime import datetime
# import httpcore
import httpx

# Parameters and setup

# set whale-alert-api-key (token)
# from key import TOKEN     # if from file
TOKEN = os.environ.get('TOKEN')

# set cursor file path
cursorpath = os.environ.get('CURSORPATH', '/data/')

# Set up logging
logging.basicConfig(format='%(asctime)s %(levelname)s: %(message)s',
                    level=logging.INFO)

# min individual transaction dollar value (free plan: 500k minimum)
min_value = os.environ.get('MIN_VALUE', 500000)
logging.info('Individual transaction minimal value: $%s', min_value)

# Elasticearch params
eshost = os.environ.get('ESHOST', 'http://elasticsearch:9200')
esindex = os.environ.get('ESINDEX', 'whalealert')
esurl = eshost + '/' + esindex + '/_doc/'
esheaders = {'Content-Type': 'application/json'}

# whale alert vars
URL = 'https://api.whale-alert.io'
TRANSACTIONURI = '/v1/transactions'
# diff = 3599


def getWhaleData(cursor):

    # timestamp = int(time.time()) - diff

    params = {'api_key': TOKEN, 'min_value': min_value, 'cursor': cursor}

    # Probably not best idea to make this call async
    # async with httpx.AsyncClient() as client:
    #     r = await client.get(url=URL + TRANSACTIONURI, params=params)
    #     return r.json()

    # sync call
    try:
        with httpx.Client() as sync_client:
            r = sync_client.get(url=URL + TRANSACTIONURI, params=params)
            return r.json()

    except Exception as e:
        logging.error('Bad response from WhaleAPI: %s', e)
        return {'result': 'failure', 'message': 'error'}

# for initial tests
# def writeToFileJson(payload):
#     with open("api-example-output.json", "w") as f:
#         f.write(json.dumps(payload, indent=4))


def removeCrusorFile():
    try:
        logging.warning('Cursor to old, trying remove cursor file...')
        os.remove(cursorpath + "cursor.json")
        return True
    except Exception as e:
        return e


def writeToFileCursor(cursor):
    with open(cursorpath + "cursor.json", "w", encoding=None) as f:
        logging.info('Updating cursor file, cursor: %s', payload['cursor'])
        cursorj = {'cursor': cursor}
        f.write(json.dumps(cursorj, indent=4))


def fetchLatestCursor():
    try:
        with open(cursorpath + "cursor.json", "r") as f:
            cursor = json.load(f)['cursor']
    except FileNotFoundError:
        # if no file, return empty cursor
        logging.info('No cursor file found, returning empty cursor')
        return '0-0-0'
    else:
        logging.info('Found recent cursor in file, cursor: %s', cursor)
        return cursor


async def postEs(client, item):
    try:
        r = await client.post(esurl, data=json.dumps(item), headers=esheaders)
        return r
    except Exception as e:
        return e


async def elasticsearchSend(payload):
    logging.info('Posting whale alert data to Elasticsearch...')

    async with httpx.AsyncClient() as client:

        start_time = time.time()
        tasks = []
        for item in payload['transactions']:

            esdate = datetime.utcfromtimestamp(
                item['timestamp']).strftime('%Y-%m-%dT%H:%M:%S')

            item['@timestamp'] = esdate

            # Post to Elasticsearch async
            tasks.append(asyncio.ensure_future(postEs(client, item)))

        responses = await asyncio.gather(*tasks)

        logging.info("--- Async total time: %s sec ---",
                     (time.time() - start_time))
        logging.info('Posting to ES finished')

        return responses


# Try to fetch old cursor from file at startup
cursor = fetchLatestCursor()

while True:
    # payload = asyncio.run(getWhaleData(cursor))
    payload = getWhaleData(cursor)

    if payload and payload['result'] == 'success':
        # save new cursor
        cursor = payload['cursor']

        if payload['count'] > 0:
            logging.info('Got new transactions! Trans count: %s',
                         payload['count'])

            r = asyncio.run(elasticsearchSend(payload))

            # optional logging, of async post operations, remove if you want
            for call in r:
                logging.info(call)

            # update new cursor in file for restarts
            # only if new data, otherwise cursor will be unchanged
            writeToFileCursor(cursor)

        else:
            logging.info('No transactions...')
            logging.info('Unchanged cursor: %s', payload['cursor'])

    else:
        logging.error(payload)
        if '3600' in payload['message']:
            cursorRm = removeCrusorFile()
            logging.info('Old cursor removed: %s', cursorRm)

    # wait X sec and run again
    # For the free plan the number of requests is limited to 10 per minute.
    time.sleep(120)
